### PR TITLE
Water/LiquidOperatingCondition interface

### PR DESCRIPTION
**Pull request by benflexcompute**

### CODE REVIEW

File: flow360/component/simulation/operating_condition/operating_condition.py [hunk 1]
```diff
@@ -426,9 +431,33 @@ def flow360_reynolds_number(self, length_unit: LengthType.Positive):
         ).value
 
 
+class LiquidOperatingCondition(Flow360BaseModel):
+    type_name: Literal["LiquidOperatingCondition"] = pd.Field(
+        "LiquidOperatingCondition", frozen=True
+    )
+    velocity_magnitude: Optional[VelocityType.Positive] = ConditionalField(
```
Comment (yifan-flex) [inline flow360/component/simulation/operating_condition/operating_condition.py:438]: Is this NonNegative?

File: flow360/component/simulation/models/surface_models.py [hunk 2]
```diff
@@ -338,6 +341,35 @@ def check_wall_function_conflict(self):
...
+    def _disable_expression_for_liquid(cls, value):
+        validation_info = get_validation_info()
+        if validation_info is None or validation_info.using_water_as_material is False:
+            return value
+
+        if isinstance(value, tuple):
+            if (
+                isinstance(value[0], str)
+                and isinstance(value[1], str)
+                and isinstance(value[2], str)
```
Comment (angranl-flex) [inline flow360/component/simulation/models/surface_models.py:366]: why not check for stringExpression like in Freestream's validator?

Comment (benflexcompute) [reply flow360/component/simulation/models/surface_models.py:366]: Oh the freestream one does not work actually.

Comment (benflexcompute) [reply flow360/component/simulation/models/surface_models.py:366]: Let me fix that. String is fine until we change how expressions are handeled.

File: flow360/component/simulation/translator/solver_translator.py [hunk 3]
```diff
@@ -905,16 +909,27 @@ def get_solver_json(
...
     op_acoustic_to_static_pressure_ratio = (
-        op.thermal_state.density * op.thermal_state.speed_of_sound**2 / op.thermal_state.pressure
-    ).value
+        (
+            op.thermal_state.density
+            * op.thermal_state.speed_of_sound**2
+            / op.thermal_state.pressure
+        ).value
+        if not isinstance(op, LiquidOperatingCondition)
+        else 1.0
```
Comment (angranl-flex) [inline flow360/component/simulation/translator/solver_translator.py:931]: Should we use 1.0 here for the LiquidOP or it does not matter?

File: tests/simulation/params/test_simulation_params.py [hunk 4]
```diff
@@ -357,3 +402,40 @@ def test_delta_temperature_scaling():
...
+        converted.models[1].velocity.value[0],
+        1.0 / fake_water_speed_of_sound,
+    )
+
+    # AngularVelocityType
+    assertions.assertAlmostEqual(
+        converted.models[2].spec.value.value,
+        0.45 / fake_water_speed_of_sound.value,
+        # Note: We did not get original value from params like the others b.c
+        # for some unknown reason THIS value in params will also become in flow360 unit system...
```
Comment (angranl-flex) [inline tests/simulation/params/test_simulation_params.py:433]: Incomplete sentence? also become "what" in unit system?

File: tests/simulation/params/test_simulation_params.py [hunk 5]
```diff
@@ -357,3 +402,40 @@ def test_delta_temperature_scaling():
...
+    )
+    # ViscosityType
+    assertions.assertAlmostEqual(
+        converted.models[3].turbulence_quantities.modified_turbulent_viscosity.value,
+        10 / (1000 * fake_water_speed_of_sound.value),
+        # Note: We did not get original value from params like the others b.c
+        # for some unknown reason THIS value in params will also become in flow360 unit system...
```
Comment (angranl-flex) [inline tests/simulation/params/test_simulation_params.py:440]: same here

File: tests/simulation/translator/test_solver_translator.py [hunk 6]
```diff
@@ -552,3 +552,7 @@ def test_boundaries():
             ],
         )
     translate_and_compare(param, mesh_unit=1 * u.m, ref_json_file="Flow360_boundaries.json")
+
+
+# def test_liquid_simulation_translation():
+#     # Note: Maybe having this as one of the local test input?
```
Comment (angranl-flex) [inline tests/simulation/translator/test_solver_translator.py:558]: Will we add the test in this or another PR?

File: tests/simulation/params/test_simulation_params.py (hunk 4 above)
Comment (benflexcompute) [reply tests/simulation/params/test_simulation_params.py:433]: It does not matter I think. It does not affect allowed models.

File: flow360/component/simulation/translator/solver_translator.py (hunk 3 above)
Comment (benflexcompute) [reply flow360/component/simulation/translator/solver_translator.py:931]: if not isinstance(op, LiquidOperatingCondition)

Comment (benflexcompute) [reply flow360/component/simulation/translator/solver_translator.py:931]: I think it does not matter. It won't affect allowed models.

File: tests/simulation/params/test_simulation_params.py (hunk 4 above)
Comment (benflexcompute) [reply tests/simulation/params/test_simulation_params.py:433]: changed

File: tests/simulation/params/test_simulation_params.py (hunk 5 above)
Comment (benflexcompute) [reply tests/simulation/params/test_simulation_params.py:440]: changed

File: tests/simulation/translator/test_solver_translator.py (hunk 6 above)
Comment (benflexcompute) [reply tests/simulation/translator/test_solver_translator.py:558]: Good point. We will.

Comment (angranl-flex) [reply tests/simulation/translator/test_solver_translator.py:558]: OK!

File: flow360/component/simulation/models/surface_models.py [hunk 7]
```diff
@@ -371,6 +374,35 @@ def check_wall_function_conflict(self):
             )
         return self
 
+    @pd.field_validator("heat_spec", mode="after")
+    @classmethod
+    def _ensure_adiabatic_wall_for_liquid(cls, value):
+        """Only allow adiabatic wall when liquid operating condition is used"""
```
Comment (benflexcompute) [inline flow360/component/simulation/models/surface_models.py:380]:
```suggestion
        """Allow only adiabatic wall when liquid operating condition is used"""
```

File: flow360/component/simulation/translator/solver_translator.py [hunk 8]
```diff
@@ -1293,4 +1316,16 @@ def get_solver_json(
                 udd_dict_translated["outputTargetName"] = udd.output_target.full_name
             translated["userDefinedDynamics"].append(udd_dict_translated)
 
+    translated["usingLiquidAsMaterial"] = isinstance(
+        input_params.operating_condition, LiquidOperatingCondition
+    )
+    translated["outputRescale"] = {}
```
Comment (benflexcompute) [inline flow360/component/simulation/translator/solver_translator.py:1322]:
```suggestion
    translated["outputRescale"] = {"velocityScale":1.0}
```

File: tests/simulation/params/test_simulation_params.py [hunk 9]
```diff
@@ -52,7 +55,11 @@
 )
 from flow360.component.simulation.simulation_params import SimulationParams
 from flow360.component.simulation.time_stepping.time_stepping import Unsteady
-from flow360.component.simulation.unit_system import CGS_unit_system, SI_unit_system
+from flow360.component.simulation.unit_system import (
+    CGS_unit_system,
+    LengthType,
```
Comment (benflexcompute) [inline tests/simulation/params/test_simulation_params.py:60]:
```suggestion
```

File: tests/simulation/translator/test_solver_translator.py [hunk 10]
```diff
@@ -186,6 +188,7 @@ def translate_and_compare(
     param, mesh_unit, ref_json_file: str, atol=1e-15, rtol=1e-10, debug=False
 ):
     translated = get_solver_json(param, mesh_unit=mesh_unit)
+    print(translated)
```
Comment (benflexcompute) [inline tests/simulation/translator/test_solver_translator.py:191]: remove

File: poetry.lock [hunk 11]
```diff
@@ -4555,7 +4555,6 @@ files = [
     {file = "ruamel.yaml.clib-0.2.12-cp310-cp310-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl", hash = "sha256:f66efbc1caa63c088dead1c4170d148eabc9b80d95fb75b6c92ac0aad2437d76"},
     {file = "ruamel.yaml.clib-0.2.12-cp310-cp310-musllinux_1_1_i686.whl", hash = "sha256:22353049ba4181685023b25b5b51a574bce33e7f51c759371a7422dcae5402a6"},
     {file = "ruamel.yaml.clib-0.2.12-cp310-cp310-musllinux_1_1_x86_64.whl", hash = "sha256:932205970b9f9991b34f55136be327501903f7c66830e9760a8ffb15b07f05cd"},
```
Comment (benflexcompute) [inline poetry.lock:4557]: why we have lock change?

File: tests/simulation/params/test_validators_params.py [hunk 12]
```diff
@@ -1424,6 +1428,69 @@ def test_deleted_surfaces():
...
+                    )
+                ),
+                PorousMedium(
+                    volumes=[porous_zone],
+                    darcy_coefficient=(0.1, 2, 1.0) / u.cm / u.m,
+                    forchheimer_coefficient=(0.1, 2, 1.0) / u.ft,
+                    volumetric_heat_source=123 * u.lb / u.s**3 / u.ft,
+                ),
+                Wall(
+                    heat_spec=HeatFlux(value=10 * u.W / u.m**2),
```
Comment (benflexcompute) [inline tests/simulation/params/test_validators_params.py:1453]: heat_spec error is missing... is it overshadowed by other error?

File: tests/simulation/params/test_validators_params.py [hunk 13]
```diff
@@ -1424,6 +1428,69 @@ def test_deleted_surfaces():
...
+                    surfaces=all_boundaries[0:-1],
+                    velocity=["1", "2", "2"],
+                ),
+                Outflow(entities=all_boundaries[-1], spec=Pressure(value=1.01e6 * u.Pa)),
```
Comment (benflexcompute) [inline tests/simulation/params/test_validators_params.py:1457]: I think we are missing checks for
 rotation-> angle expression error
 boundary condition type selection (cannot use Inflow for exmple)

can you add them?

File: tests/simulation/params/test_validators_params.py (hunk 12 above)
Comment (angranl-flex) [reply tests/simulation/params/test_validators_params.py:1457]: Added the assertion check

**Review by angranl-flex — APPROVED** [review]
